core-libs/setup/ssr/engine/cx-common-engine-wrapper.ts

```diff
@@ -0,0 +1,54 @@
+/*
+ * SPDX-FileCopyrightText: 2024 SAP Spartacus team <spartacus-team@sap.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { StaticProvider } from '@angular/core';
+import { CommonEngine } from '@angular/ssr';
+import { CxServerErrorResponse } from '../error-handling/server-error-response';
+import { PROPAGATE_SERVER_ERROR_RESPONSE } from '../error-handling/server-error-response/propagate-server-error-response';
+import { RenderOptions } from './ng-express-engine';
+
+/**
+ * Wraps the CommonEngine render method to handle propagated server responses caught during server-side rendering.
+ */
+
+export function cxCommonEngineWrapper(
+  engine: CommonEngine,
+  renderOptions: RenderOptions,
+  callback: (err?: Error | null, html?: string) => void
+) {
+  const propagatedServerErrorResponses: (CxServerErrorResponse | string)[] = [];
```

Comment:
We don't need to store an array (with mixed errors and string htmls), but we need to store just one potential element `errorResponse`.

What about simply having a single variable instead of array
```diff
- const propagatedServerErrorResponses: (CxServerErrorResponse | string)[] = [];
+ `let errorResponse: undefined | CxServerResponse`.
```

and
```diff
-       return (serverErrorResponse: CxServerErrorResponse) => {
-         propagatedServerErrorResponses.push(serverErrorResponse);
-      };
-       return (serverErrorResponse: CxServerErrorResponse) => {
-         errorResponse ??= serverErrorResponse
-      };
```

```diff
- .then((html) => propagatedServerErrorResponses.push(html))
+ .then((html) => {
+    if (errorResponse) { 
+      throw errorResponse;
+    }
+    return html
+  })
``` 

No need to catch errors from APP_INITIALIZERs. If Angular throws anytime, we should let it bubble up the promises chain immediately
```diff
- .catch((err) => propagatedServerErrorResponses.push(err))
``` 

And no need for `finally` IMHO. We've already handled 2 cases:
- on success, we'll either return the produced html or throw the stored `errorResponse` (if exists)
- on error (APP_INITIALIZER only), we've already bubbled the error up.